File: 1.Bronze/3.Simulation/art.cpp
#include <iostream>
#include <fstream>
#include <vector>

using std::cout;
using std::endl;
using std::vector;

const int MAX_COLOR = 9;

int main() {
    /*
     * Bounding boxes of each of the colors.
     * The first element won't be used (colors go from 1-9)
     */
    vector<int> left(MAX_COLOR + 1);
    vector<int> right(MAX_COLOR + 1);
    vector<int> up(MAX_COLOR + 1);
    vector<int> down(MAX_COLOR + 1);
    for (int c = 1; c <= MAX_COLOR; c++) {
        left[c] = up[c] = INT32_MAX;
        right[c] = down[c] = -1;
    }

    vector<bool> valid_start(MAX_COLOR + 1);
    std::ifstream read("art.in");
    int width;
    read >> width;
    vector<vector<int>> art(width, vector<int>(width));
    for (int r = 0; r < width; r++) {
        for (int c = 0; c < width; c++) {
            char curr_char;
            read >> curr_char;
            int curr = curr_char - '0';
            art[r][c] = curr;
            if (curr != 0) {
                left[curr] = std::min(left[curr], c);
                right[curr] = std::max(right[curr], c);
                down[curr] = std::max(down[curr], r);
                up[curr] = std::min(up[curr], r);
                valid_start[curr] = true;
            }
        }
    }

    for (int color = 1; color <= MAX_COLOR; color++) {
        for (int r = up[color]; r <= down[color]; r++) {
            for (int c = left[color]; c <= right[color]; c++) {
                if (art[r][c] != color) {
                    valid_start[art[r][c]] = false;
                }
            }
        }
    }

    int total_starts = 0;
    for (bool s : valid_start) {
        total_starts += s ? 1 : 0;
    }
    std::ofstream("art.out") << total_starts << endl;
}

File: A.Bronze/4.Simulation/art.java
import java.io.*;

public class art {
    static final int MAX_COLOR = 9;

    public static void main(String[] args) throws IOException {
        BufferedReader read = new BufferedReader(new FileReader("art.in"));
        int width = Integer.parseInt(read.readLine());

        /*
         * Bounding boxes of each of the colors.
         * The first element won't be used (colors go from 1-9)
         */
        int[] left = new int[MAX_COLOR + 1];
        int[] right = new int[MAX_COLOR + 1];
        int[] down = new int[MAX_COLOR + 1];
        int[] up = new int[MAX_COLOR + 1];
        for (int c = 1; c <= MAX_COLOR; c++) {
            left[c] = up[c] = Integer.MAX_VALUE;
            right[c] = down[c] = -1;
        }
        
        boolean[] validStart = new boolean[MAX_COLOR + 1];
        int[][] art = new int[width][width];
        for (int r = 0; r < width; r++) {
            String row = read.readLine();
            for (int c = 0; c < width; c++) {
                int curr = Character.getNumericValue(row.charAt(c));
                art[r][c] = curr;
                if (curr != 0) {
                    left[curr] = Math.min(left[curr], c);
                    right[curr] = Math.max(right[curr], c);
                    down[curr] = Math.max(down[curr], r);
                    up[curr] = Math.min(up[curr], r);
                    validStart[curr] = true;
                }
            }
        }

        for (int color = 1; color <= MAX_COLOR; color++) {
            for (int r = up[color]; r <= down[color]; r++) {
                for (int c = left[color]; c <= right[color]; c++) {
                    if (art[r][c] != color) {
                        validStart[art[r][c]] = false;
                    }
                }
            }
        }

        int totalStarts = 0;
        for (boolean c : validStart) {
            totalStarts += c ? 1 : 0;
        }

        PrintWriter written = new PrintWriter("art.out");
        written.print(totalStarts);
        written.close();
    }
}